feat: add DisjointIntervalSequence API (#210)
deepgenomics/GenomeKit: docs-src/diseq.rst

Reviewer: I find this writing style a little too verbose and personal, let's discuss.

.. _diseq:

-------------------------------
Disjoint Interval Sequences
-------------------------------

Motivation
==========

When working with transcripts, the situation may arise where we want to ignore the
introns (or other features) of the transcript. If you were to represent the transcript
with those parts removed, you would be left with a disjoint series of :py:class:`~genome_kit.Interval`
objects that are difficult to work with directly (for example, creating an interval on
this disjoint space, or querying the position of a specific sequence within a
CDS, relative to the spliced RNA sequence).

For this reason, the :py:class:`~genome_kit.diseq.DisjointIntervalSequence` class was
introduced. :py:class:`~genome_kit.diseq.DisjointIntervalSequence` (DIS) simplifies
working with intervals that exist on a disjoint coordinate space.

Overview
========

A :py:class:`~genome_kit.diseq.DisjointIntervalSequence` (DIS) represents
a flattened coordinate system over a sequence of disjoint genomic intervals.
For example, the exons of a transcript form a disjoint set of genomic intervals that,
when concatenated, represent the spliced RNA sequence.

> A DIS has two layers:

Reviewer: ```suggestion
A DIS has two aspects:
```

In what way does the layer metaphor work?

Author: The thinking was you have the coordinate space (layer 1), and then the interval (layer 2) that exists upon the coordinate (layer 1). Ultimately it's just semantics, and I'm not opposed to changing the wording. Accepted your suggestion.

                          end5      Interval      end3
    Start Index:     0
    End Index:       7

We see that the interval spans the full length of the coordinate space, and is defined by
a start index of 0 and an end index of 7.

.. note::
    Why 7 and not 6? The disjoint interval follows the convention of
    :py:class:`~genome_kit.Interval` where intervals are half-open
    (the end index is exclusive).

The above example illustrates the basics of how a DIS works. However, it is possible to
do more. We can instead define an interval within the DIS on the strand opposite that 
of the coordinate space. Let's start with the DIS coordinate space from above
::
    DIS Coordinates:       0   1   2   3   4   5   6   7
    DNA Sequence:          A   T   G   G   C   A   T
                           |<----->|   |<->|   |<->|
                             Exon1     Exon2   Exon3

Now let's add the negative (opposite) strand to the diagram
::
                                Plus Strand
    DIS Coordinates:       0   1   2   3   4   5   6   7
    DNA Sequence (+):      A   T   G   G   C   A   T
    -----------------------------------------------------

>     DNA Sequence (-):      T   A   C   C   G   T   A

Reviewer: It's a little unforunate that the complement in reverse order is so similar to the original:
`ATGGCAT` vs `ATGCCAT`.

Can you change the sequence to make the examples potentially less confusing?

    DIS Coordinates:       0   1   2   3   4   5   6   7
                                Minus Strand

Importantly, we see the DIS coordinates are the same on both strands. This simplifies
things when you want to get the complement of a given interval, as you can use the same
indices and just flip the strand. To illustrate this, let's now define the same interval
as before (spanning the entire coordinate space) but on the negative strand
::
                                Plus Strand
    DIS Coordinates:       0   1   2   3   4   5   6   7
    DNA Sequence (+):      A   T   G   G   C   A   T
    -----------------------------------------------------
    DNA Sequence (-):      T   A   C   C   G   T   A
    DIS Coordinates:       0   1   2   3   4   5   6   7
                                Minus Strand
                           |<--------------------->|
                          end3      Interval      end5
    Start Index:     0
    End Index:       7
    On Coordinate Strand: False

You will notice "On Coordinate Strand: False" has been added to the diagram. Since we
aren't able to determine which strand the interval is on just from the indices, this

> variable is used to let us know the strandedness of the interval.

Reviewer: I don't really get this. Does "On Coordinate Strand" map to a field on the DSI? Or are you just trying to explain the diagram?

Author: It does map to a field on the DIS. I'll rework that section to include code as well, which should hopefully clear things up a bit.

Thus far we have only defined a DIS from intervals on the + strand. When defining
a DIS from intervals on the negative strand, much remains the same. However, there is
one important difference from a regular Interval: On a DIS created from negative-strand
intervals, the indices still increase in the 5'→3' direction of the transcript. Let's

> take a look at an example:

Reviewer: Isn't it sufficient to say that this behaves the same as a vanilla interval?

Author: Vanilla Intervals actually have the indices run 3' --> 5' on the negative strand. Given this breaks that pattern, I think it's worth explaining.

Reviewer: Can we not avoid breaking the pattern?

Author: Discussed offline: we aren't exactly breaking a pattern since DIS is philosophically/conceptually different from `Interval` (i.e. DIS is not just an Interval on a disjoint coordinate). A DIS is meant to represent a spliced transcript, which has no strand, and is on a new coordinate-space separate from the genomic coordinate space. For these reasons, running 5' -> 3' and dropping notions of +/- strand (replacing with same/opposite-strand) are fine, and preferred since they more closely align with how one typically reasons about transcripts/spliced RNA.

Say we have the following transcript on the negative strand represented by the diagram
below:
::

>     Reminder: On the - strand, the 3' end is on the left and the 5' end is on the right!

Reviewer: Instead of this, mark the 3' and 5' ends on the diagram

                             Exon3      Intron2    Exon2      Intron1      Exon1
                          |<------->| |<------->| |<--->| |<----------->| |<--->|
    DNA Sequence (-):      G   T   C   A   G   T   C   A   G   T   C   A   G   T
    Genomic Coordinates:  153 154 155 156 157 158 159 160 161 162 163 164 165 166 167

>                                             Minus Strand

Reviewer: ```suggestion
                                            Negative Strand
```

(and elsewhere)

    DNA Sequence:          T   G   A   C   C   T   G
                           |<--------------------->|
                          end5      Interval      end3
    Start Index:     0
    End Index:       7
    On Coordinate Strand: True

We see that despite creating the DIS from the negative strand, the full-length interval
on the coordinate strand still looks the same as in the + strand example. When working
with DIS objects, you only need to think of things in terms of "same strand" or
"opposite strand".

To complete the example, let's define an interval on this DIS that is on the opposite strand of the coordinate space
::
    DIS Coordinates:       0   1   2   3   4   5   6   7
    DNA Sequence (-):      T   G   A   C   C   T   G
    -----------------------------------------------------
    DNA Sequence (+):      A   C   T   G   G   A   C
    DIS Coordinates:       0   1   2   3   4   5   6   7
                           |<--------------------->|
                          end3      Interval      end5
    Start Index:     0
    End Index:       7
    On Coordinate Strand: False

Now that you understand how a DIS works conceptually, you can read on to see how to

> manipulate them in code.

Reviewer: Actually I think the examples are clearer when shown together with code examples. e.g "despite creating the DIS from the negative strand" is more difficult to grasp in isolation.

Construction
============

From a Transcript
~~~~~~~~~~~~~~~~~

The most common way to create a DIS is from a
:py:class:`~genome_kit.Transcript`::

    >>> from genome_kit import Genome
    >>> from genome_kit.diseq import DisjointIntervalSequence
    >>> genome = Genome("gencode.v29")
    >>> transcript = genome.transcripts[100]
    >>> dis = DisjointIntervalSequence.from_transcript(transcript)

By default, the coordinate space is built from the transcript's exons.
You can also specify a region to use CDS or UTR intervals::

    >>> dis_cds = DisjointIntervalSequence.from_transcript(transcript, region="cds")
    >>> dis_utr5 = DisjointIntervalSequence.from_transcript(transcript, region="utr5")
    >>> dis_utr3 = DisjointIntervalSequence.from_transcript(transcript, region="utr3")

> The ``coord_id`` and ``interval_id`` default to ``transcript.id`` but can

Reviewer: `coord_id` and `interval_id` are introduced without explaining what they are or what purpose they serve.

Author: They are just intended to be labels/names used by users. Changed var names from ID to name to clear this up.

Reviewer: why are they needed?

be overridden::

    >>> dis = DisjointIntervalSequence.from_transcript(
    ...     transcript, coord_id="my_coord", interval_id="my_interval")

From Intervals
~~~~~~~~~~~~~~

You can also construct a DIS from any sequence of
:py:class:`~genome_kit.Interval` objects (or annotation objects like

> :py:class:`~genome_kit.Exon` that have an ``.interval`` attribute)::

Reviewer: 1. isn't this true for any Exon instance?
2. Exon inherits from Interval, so this distinction isn't necessary.

    >>> from genome_kit import Interval
    >>> exon_intervals = [e.interval for e in transcript.exons]
    >>> dis = DisjointIntervalSequence.from_intervals(exon_intervals, coord_id="my_coord")

The intervals must all share the same chromosome, strand, and reference

> genome. They are automatically sorted 5'→3' and checked for overlaps.

Reviewer: ```suggestion
The intervals must all share the same chromosome, strand, and reference
genome and must not overlap. They are automatically sorted 5'→3'.
```

    0
    >>> dis.end
    250
    >>> dis.length
    250
    >>> len(dis)
    250

By default, the interval spans the full coordinate space (``start=0``,
``end=coordinate_length``). Indices can extend beyond ``[0, coordinate_length]``, but
the DNA sequence returned by ``genome.dna()`` will be N-padded.

End5 and End3
~~~~~~~~~~~~~

The ``end5_index`` and ``end3_index`` properties give the 5' and 3' positions
of the interval. These are derived from ``start`` and ``end`` based on the
interval's strand::
    On coordinate strand (on_coordinate_strand=True):
    Start Index:     1
    End Index:       6
    DIS Coordinates:       0   1   2   3   4   5   6   7
    DNA Sequence:          T   G   A   C   C   T   G
                               |<------------->|
    -----------------------------------------------------
    DNA Sequence:          A   C   T   G   G   A   C
    DIS Coordinates:       0   1   2   3   4   5   6   7
                                  Opposite Strand

    >>> dis = DisjointIntervalSequence.from_transcript(transcript)
    >>> dis.end5_index   # same as start when on coordinate strand
    1
    >>> dis.end3_index   # same as end when on coordinate strand
    6

>     Off coordinate strand (on_coordinate_strand=False):

Reviewer: What's the use case?

Author: Basically some of the user feedback I got was they wanted to have an interval on the opposite strand as the coordinate, but maintain the same indices to keep things easy to reason about. One use-case was to get the complement of a sequence, or to see if a given sequence would be on the opposite strand of the DIS.

Reviewer: Not crazy about this since it adds complexity and breaks compat with intervals. Can we provide a convenience method (or methods?) instead, to give the users the behavior they need?

Author: Added a comment in [another thread](https://github.com/deepgenomics/GenomeKit/pull/210#discussion_r3059326131), but this doesn't break the pattern since DIS is not meant to represent/act like `Interval`

    Start Index:     3
    End Index:       7
    DIS Coordinates:       0   1   2   3   4   5   6   7
    DNA Sequence:          T   G   A   C   C   T   G
    -----------------------------------------------------
    DNA Sequence:          A   C   T   G   G   A   C
    DIS Coordinates:       0   1   2   3   4   5   6   7
                                       |<--------->|
                                  Opposite Strand

    >>> opp = dis.as_opposite_strand()
    >>> opp.end5_index   # same as end when off coordinate strand
    7
    >>> opp.end3_index   # same as start when off coordinate strand

>     3

Author: This will be added in the next PR. IMO, it's fine as is, but open to input from others.

    -----------------------------------------------------
    DNA Sequence (-):      T   A   C   G   T   C   G
    DIS Coordinates:       0   1   2   3   4   5   6   7
                                  Opposite Strand
                           |<--------------------->|
                          end3      Interval      end5
    Start Index:     0
    End Index:       7
    On Coordinate Strand: False

The ``on_coordinate_strand`` flag distinguishes same-strand from opposite-strand
intervals, since the start and end indices alone do not encode strand information::

    >>> dis_opp = DisjointIntervalSequence(
    ...     [exon1, exon2, exon3],
    ...     coord_name="tx_example",
    ...     on_coordinate_strand=False,
    ... )
    >>> dis_opp.on_coordinate_strand
    False
    >>> dis_opp.end5_index
    7
    >>> dis_opp.end3_index
    0

The preceding examples used + strand coordinate intervals. When the coordinate intervals
lie on the - strand, the DIS behaves identically in most respects, with one key
difference from :py:class:`~genome_kit.Interval`: on a DIS created from
negative-strand intervals, indices still increase in the 5'→3' direction of the

> transcript.

Reviewer: not blocking, but a code example that demonstrates the behavioral differences between a vanilla interval and a simple dji interval would be helpful.
